### ISSUE TITLE
When comparing amounts in different commodities, commodity symbol is compared first, even when amounts have different signs

### ISSUE THREAD

**Issue by Xitian9** [simonmichael/hledger#1563]
The problem of sorting amounts expressed in different commodities is a tough one. Do you convert them to some reference commodity when there's price information, or just compare as is.

Though there's certainly a more detailed conversation to have on this topic, I think there's one aspect of the current behaviour which can clearly be improved: negative numbers in one commodity should always less than positive numbers in another commodity. hledger does not currently respect this: see the following example.

```
$ cat sort.j
P 2021-01-01 YYY XXX1

2021-01-01 Post in XXX
    (a)   100 XXX

2021-01-01 Post in XXX
    (b)  -100 XXX

2021-01-02 Post in XXX
    (c)   100 XXX

2021-01-02 Post in YYY
    (d)  -100 YYY

$  hledger -f sort.j bal -DN --sort -V
Balance changes in 2021-01-01..2021-01-02, valued at period ends:

   || 2021-01-01  2021-01-02 
===++========================
 a ||    100 XXX           0 
 c ||          0     100 XXX 
 b ||   -100 XXX           0 
 d ||          0    -100 XXX 


$  hledger -f sort.j bal -DN --sort
Balance changes in 2021-01-01..2021-01-02:

   || 2021-01-01  2021-01-02 
===++========================
 d ||          0    -100 YYY 
 a ||    100 XXX           0 
 c ||          0     100 XXX 
 b ||   -100 XXX           0

**Comment by Xitian9** [comment]
That said, it looks like the simplest way to solve this problem involves coming up with a better `Ord` instance for `MixedAmount`, so we may have to address the more general problem. My proposal for comparing two mixed amounts:

1. Strip prices.
1. Convert each to their list of `Amount`s.
2. Sort those amounts by `aquantity`.
3. Compare the resulting lists.

It's not perfect, but should perform reasonably in most common cases.

**Comment by simonmichael** [comment]
Some more testing to help me understand this. Using your example, only slightly simplified.:
```
$ hledger bal -N  # Sorting by account is the default
               100 X  a
              -100 X  b
               100 X  c
              -100 Y  d
$ hledger bal -N --sort  # 1.21's sorting by amount. Hard to see the logic.
              -100 Y  d
               100 X  a
               100 X  c
              -100 X  b
$ hledger.1564 bal -N --sort  # #1564's sorting by amount. Numbers look a bit more sorted, at least in this example.
               100 X  a
               100 X  c
              -100 Y  d
              -100 X  b
# Explanation of 1564's sort: it treats these amounts as
               100 X,    0 Y  a
               100 X,    0 Y  c
                 0 X, -100 Y  d
              -100 X,    0 Y  b
# X is alphabetically before Y, so the X amount determines the order. If X amounts are equal, the Y amount will decide, and so on.
```
It's the same with the multi-period report, with the Total column controlling the sort (whether made visible with -T or not).

 When all commodity amounts are equal (a and c above), their sort order is what - undefined ?

**Comment by simonmichael** [comment]
Fixed by #1564.
